go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -391,3 +668,526 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+		return newVCopierCopyWorker(
+			false, /* close db client */
+			vc.vr.dbClient,
+		), nil
+	}
+	if parallelism > 1 {
+		workerFactory = func(ctx context.Context) (*vcopierCopyWorker, error) {
+			dbClient, err := vc.newClientConnection(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create new db client: %s", err.Error())
+			}
+			return newVCopierCopyWorker(
+				true, /* close db client */
+				dbClient,
+			), nil
+		}
+	}
+	return workerFactory
+}
+
+// close waits for all workers to be returned to the worker pool.
+func (vcq *vcopierCopyWorkQueue) close() {
+	if !vcq.isOpen {
+		return
+	}
+	vcq.isOpen = false
+	vcq.workerPool.Close()
+}
+
+// enqueue a new copy task. This will obtain a worker from the pool, execute
+// the task with that worker, and afterwards return the worker to the pool. If
+// vcopierCopyWorkQueue is configured to operate concurrently, the task will be
+// executed in a separate goroutine. Otherwise the task will be executed in the
+// calling thread.
+func (vcq *vcopierCopyWorkQueue) enqueue(ctx context.Context, currT *vcopierCopyTask) error {
+	if !vcq.isOpen {
+		return fmt.Errorf("work queue is not open")
+	}
+
+	// Get a handle on an unused worker.
+	poolH, err := vcq.workerPool.Get(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to get a worker from pool: %s", err.Error())
+	}
+
+	currW, ok := poolH.(*vcopierCopyWorker)
+	if !ok {
+		return fmt.Errorf("failed to cast pool resource to *vcopierCopyWorker")
+	}
+
+	execute := func(task *vcopierCopyTask) {
+		currW.execute(ctx, task)
+		vcq.workerPool.Put(poolH)
+	}
+
+	// If the work queue is configured to work concurrently, execute the task
+	// in a separate goroutine. Otherwise execute the task in the calling
+	// goroutine.
+	if vcq.concurrent {
+		go execute(currT)
+	} else {
+		execute(currT)
+	}
+
+	return nil
+}
+
+// open the work queue. The provided arguments are used to generate
+// statements for inserting rows and copy state.
+func (vcq *vcopierCopyWorkQueue) open(
+	copyStateInsert *sqlparser.ParsedQuery,
+	pkfields []*querypb.Field,
+	tablePlan *TablePlan,
+) {
+	if vcq.isOpen {
+		return
+	}
+
+	poolCapacity := int(math.Max(float64(vcq.maxDepth), 1))
+	vcq.workerPool = pools.NewResourcePool(
+		/* factory */
+		func(ctx context.Context) (pools.Resource, error) {
+			worker, err := vcq.workerFactory(ctx)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to create vcopier worker: %s",
+					err.Error(),
+				)
+			}
+			worker.open(copyStateInsert, pkfields, tablePlan)
+			return worker, nil
+		},
+		poolCapacity, /* initial capacity */
+		poolCapacity, /* max capacity */
+		0,            /* idle timeout */
+		nil,          /* log wait */
+		nil,          /* refresh check */
+		0,            /* refresh interval */
+	)
+
+	vcq.isOpen = true
+}
+
+// after returns a vcopierCopyTaskHooks that can be used to register callbacks
+// to be triggered after the task successfully advances past to the specified
```

Comment:
I can make that change, but I think "finishes/completes" isn't totally accurate since hooks can be executed after any `vcopierCopyTaskState`, and not only when a task finishes/completes.